### PR TITLE
793: Allocate assessments

### PR DESCRIPTION
**Pull request by brownbearsoftware**
**Create and allocate an Assessment on submission of Application**
Finds the user who is an assessor, with all of the required qualifications (if any) who has the least non-submitted Assessments allocated to them.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AllocationQueryTest.kt [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserQualification
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRepository
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRole
+
+class AllocationQueryTest : IntegrationTestBase() {
+  @Autowired
+  lateinit var realUserRepository: UserRepository
+
+  @Test
+  fun `findQualifiedAssessorWithLeastPendingAllocations query works as described`() {
```
Comment (pezholio) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/integration/AllocationQueryTest.kt:15]: I think this test is good, but I'm finding it a bit hard to follow - perhaps with some `given/when/then` type comments to make it clear what is happening in the setup?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt [hunk 2]
```diff
@@ -1,11 +1,54 @@
...
+import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.JsonSchemaType
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserEntity
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserQualification
 import uk.gov.justice.digital.hmpps.approvedpremisesapi.jpa.entity.UserRepository
+import java.time.OffsetDateTime
+import java.util.UUID
 
 @Service
-class AssessmentService(private val userRepository: UserRepository) {
+class AssessmentService(
```
Comment (pezholio) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt:15]: Does this get called anywhere yet?

**Review by pezholio — APPROVED** [review]
LGTM - just a couple of questions / suggestions

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt [hunk 3]
```diff
@@ -37,7 +37,7 @@ data class AssessmentEntity(
 
   @ManyToOne
   @JoinColumn(name = "allocated_to_user_id")
-  val allocatedToUserId: UserEntity,
+  val allocatedToUser: UserEntity,
```
[the remark this answers is not in the document]
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/jpa/entity/AssessmentEntity.kt:40]: No migration needed, the database column name renames the same - just changed this as it's actually an entire User once JPA has done its magic rather than just the ID

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt (hunk 2 above)
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt:15]: So I put this in first in a commit then started using it in a later commit from ApplicationService

File: src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt [hunk 4]
```diff
@@ -1,11 +1,54 @@
...
+  private val userRepository: UserRepository,
+  private val assessmentRepository: AssessmentRepository,
+  private val jsonSchemaService: JsonSchemaService
+) {
+  fun createAssessment(application: ApplicationEntity): AssessmentEntity {
+    val requiredQualifications = getRequiredQualifications(application)
+
+    val allocatedUser = getUserForAllocation(requiredQualifications)
+      ?: throw RuntimeException("No Users with all of required qualifications (${requiredQualifications.joinToString(", ")}) could be found")
```
[the remark this answers is not in the document]
Comment (brownbearsoftware) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/AssessmentService.kt:25]: 👍
